### PR TITLE
Dat 3246 round 4

### PR DESCRIPTION
**Pull request by nikodamn**
This is fourth round of automated tests for Infobox Builder. There's also one fix for template classification test, which has been involved with introducing that feature.

**Reviewers**
@ludwikkazmierczak @tomasznapieralski @drets @justnpT

### CODE REVIEW

File: src/test/java/com/wikia/webdriver/elements/oasis/pages/WikiFeatures.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  public WikiFeatures enableEuropaInfoboxTheme() {
+    wait.forElementClickable(europaInfoboxThemeSlider);
+    europaInfoboxThemeSlider.click();
+
+    return this;
+  }
+
+  public WikiFeatures disableEuropaInfoboxTheme() {
+    if (this.isEuropaInfoboxThemeEnabled() == true) {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/elements/oasis/pages/WikiFeatures.java]: MINOR Remove the literal "true" boolean value. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1125)

File: src/test/java/com/wikia/webdriver/elements/oasis/pages/WikiFeatures.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+  public WikiFeatures openWikiFeatures(String wikiURL) {
+    getUrl(wikiURL + URLsContent.SPECIAL_WIKI_FEATURES);
+    PageObjectLogging.log("openWikiFeaturesPage", "special wiki features page opened", true, driver);
+
+    return this;
+  }
+
+  public boolean isEuropaInfoboxThemeEnabled() {
+    if (europaInfoboxThemeSliderText.getText().equalsIgnoreCase("Enabled")) {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/elements/oasis/pages/WikiFeatures.java]: MAJOR Replace this if-then-else statement by a single return statement. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1126)
MAJOR Move the "Enabled" string literal on the left side of this string comparison. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1132)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java [hunk 3]
```diff
@@ -94,7 +121,7 @@ public InfoboxBuilderPage() {
   public InfoboxBuilderPage openNew(String templateName) {
```
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: Can't we open new template by just modifying URL?

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java [hunk 4]
```diff
@@ -106,13 +133,126 @@ public InfoboxBuilderPage openExisting(String templateName) {
...
+    return modalEditTitle.isDisplayed();
+  }
+
+  public boolean isEditTemplateTitleInputPresent() {
+    wait.forElementVisible(editTemplateTitleInput);
+
+    return editTemplateTitleInput.isDisplayed();
+  }
+
+  public void insertTemplateTitle(String title) {
```
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: return `InfoboxBuilderPage`

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java [hunk 5]
```diff
@@ -1,16 +1,16 @@
-package com.wikia.webdriver.elements.oasis.pages;
+package com.wikia.webdriver.elements.mercury.pages;
 
+import com.wikia.webdriver.common.contentpatterns.URLsContent;
 import com.wikia.webdriver.common.core.Assertion;
```
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: This class grows pretty big, I would suggest to split this into components, like sidebar and main view

Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: Few unused fields in this class: 
+ helpDialog
+ successIcon
+ sectionHeadersChevron

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java [hunk 6]
```diff
@@ -136,9 +304,36 @@ public int countHeaders() {
     return headers.size();
   }
 
-  public String getBackgroundColor() {
-    wait.forElementVisible(component.get(0));
-    return component.get(0).getCssValue("background-color");
+  public InfoboxBuilderPage changeHeaderCollapsibilityState(int index) {
+    JavascriptExecutor js = (JavascriptExecutor) driver;
```
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: I think there is no need to cast driver to JS executor

File: src/test/java/com/wikia/webdriver/elements/oasis/pages/TemplateEditPage.java [hunk 7]
```diff
@@ -3,23 +3,30 @@
...
+
   public TemplateEditPage() {
     super();
   }
 
   public TemplateEditPage open(String templateName) {
     getUrl(urlBuilder.appendQueryStringToURL(String.format("%s%s%s:%s", urlBuilder.getUrlForWiki(),
-        URLsContent.WIKI_DIR, URLsContent.TEMPLATE_NAMESPACE, templateName),
-        URLsContent.ACTION_EDIT));
+                                                           URLsContent.WIKI_DIR,
```
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/oasis/pages/TemplateEditPage.java]: Seems like your formatted is a bit off, sure you are using the default wikia one?

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java (hunk 3 above)
Comment (nikodamn) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: The purpose of that is to test whole flow with it's entry point. It's navigating to template page and selecting the Infobox template type. By modifying the URL, we would test it without that important part.

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java (hunk 5 above)
Comment (nikodamn) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: We already tried to separate f.e. modal as separate component, but then we realized that with current implementation, we are unable to reuse it.

I agree with splitting that class, although I would like to do this as separate task. The reason for that is this feature needs that tests asap.

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java (hunk 6 above)
Comment (nikodamn) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: Can you explain why? It's pseudoelement on :after and my only idea how to test it was by executing js here. Or if this remark is only about casting driver to JS, please suggest how I should implement that one.

Thanks!

File: src/test/java/com/wikia/webdriver/elements/oasis/pages/TemplateEditPage.java (hunk 7 above)
Comment (nikodamn) [inline src/test/java/com/wikia/webdriver/elements/oasis/pages/TemplateEditPage.java]: Yea, I double checked. That's bizarre... Thanks for spotting it!

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java [hunk 8]
```diff
@@ -136,9 +299,36 @@ public int countHeaders() {
...
+    headers.get(index).click();
+
+    wait.forElementClickable(sidebarCheckbox);
+    sidebarCheckbox.click();
+
+    String script = "return window.getComputedStyle(document"
```
Comment (drets) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: FYI: soon we are going to merge to master the new method which could be useful here
https://github.com/Wikia/selenium-tests/pull/1120/files#diff-ceedb0f5093fdab9b86010a3e98a42b9R723

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java [hunk 9]
```diff
@@ -1,16 +1,15 @@
...
+package com.wikia.webdriver.elements.mercury.pages;
 
+import com.wikia.webdriver.common.contentpatterns.URLsContent;
 import com.wikia.webdriver.common.core.Assertion;
+import com.wikia.webdriver.elements.oasis.pages.TemplateEditPage;
 import com.wikia.webdriver.pageobjectsfactory.pageobject.special.SpecialPageObject;
-
-import org.openqa.selenium.By;
 import org.openqa.selenium.Dimension;
 import org.openqa.selenium.JavascriptExecutor;
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: MINOR Remove this unused import 'org.openqa.selenium.JavascriptExecutor'. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java (hunk 5 above)
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: Cool, remember to rise a ticket :)

File: src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java (hunk 6 above)
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/pages/InfoboxBuilderPage.java]: this line: ```JavascriptExecutor js = (JavascriptExecutor) driver;```
is not needed, `driver` object is already inheriting `JavascriptExecutor` class. Just use:
```driver.executeScript(script).toString();``` :)

**Comment by nikodamn** [comment]
@ludwikkazmierczak castint to JS removed due to your suggestions

**Comment by nikodamn** [comment]
As agreed per offline discussion that there are no other remarks and class refactoring will be provided in separate PR, I'm merging this changes.

**Pull request merged**